packages/stream-plugin/src/index.ts

```diff
@@ -44,46 +45,67 @@ export interface ActionCtx<T extends { [K in keyof T]: Stream<any> }> {
   [valueSymbol]: Partial<UnStreams<T>>;
 }
 
-const init = <T extends { [K in keyof T]: Stream<any> }>(
-  _config: {},
-  universe: Universe<T>,
-) => {
-  universe.streams = {};
-};
+export interface Event<T extends { [K in keyof T]: Stream<any> }> {
+  streams: { [K in keyof T]?: T[K] | (typeof deletedSymbol) };
+}
 
-const prepareActionCtx = <T extends { [K in keyof T]: Stream<any> }>(
-  state: State<T>,
-  streamUpdate: PluginAction<ActionCtx<T>, [keyof T, any]>,
-) => ({
+const prepareActionCtx = <T extends { [K in keyof T]: Stream<any> }>({
   ctx,
   universe,
+  event,
 }: {
-  ctx: PluginActionCtx<ActionCtx<T>, Universe<T>> & ActionCtx<T>;
+  ctx: ActionCtx<T>;
   universe: Universe<T>;
+  event: Event<T>;
 }) => {
+  // For the streamUpdate action
   ctx[valueSymbol] = universe.streams;
-  ctx.streams = new Proxy(state.streams, {
-    set: (obj, key: keyof T, value) => {
-      const cb = (value: any) => {
-        ctx.rootDispatch(streamUpdate)(key, value);
-      };
-      state.states[key] = value.subscribe(cb);
+
+  // For user actions
+  event.streams = {};
+  ctx.streams = new Proxy(event.streams as Partial<T>, {
+    get: () => undefined,
+    set: <K extends keyof T>(obj, key: K, value: T[K]) => {
       obj[key] = value;
       return true;
     },
     deleteProperty: (obj, key: keyof T) => {
-      const streamState = state.states[key];
-      if (streamState != null) {
-        streamState.unsubscribe();
-      }
+      // Typescript doesn't know about proxies (typescript#20846)
+      obj[key] = deletedSymbol as any;
       delete universe.streams[key];
-      delete state.states[key];
-      delete obj[key];
       return true;
     },
   });
 };
 
+const onCompleteEvent = <T extends { [K in keyof T]: Stream<any> }>(
+  state: State<T>,
+  streamUpdate: PluginAction<ActionCtx<T>, [keyof T, any]>,
+): PluginOnCompleteEventFn<{}, Event<T>, ActionCtx<T>> => ({
+  event,
+  rootDispatch,
+}) => {
+  // Not sure how to type the Object.entries version
+  (Object.keys(event.streams) as Array<keyof T>).forEach(key => {
+    const stream = event.streams[key];
+
+    // First, delete the old stream
+    const streamState = state.states[key];
+    if (streamState != null) {
+      streamState.unsubscribe();
+    }
+
+    if (stream === deletedSymbol) {
```

Comment:
From the above comment, it seems like there should be another explaining what this bit of code does.

Reply:
I have rephrased that, it was supposed to make clear that I need to delete the old stream even if I'm adding a new one (and it's not a bug that should be moved to the `deletedSymbol` case).

Reply:
I have rephrased the other comment, just wanted to make clear the old stream needs to be deleted in all cases, not just when a stream is deleted.

Reply:
I've rephrased the other comment. I meant to make clear that the stream always needs to be deleted, rather than just in the `deleteSymbol` case.